manager/controllers/app/moduleinstance.go

```diff
@@ -53,43 +66,84 @@ Updates to add ingest:
    - Dependencies are checked but not added yet to the blueprint
 */
 
-// GetCopyDestination chooses one of the buckets pre-allocated for use by implicit copies or ingest.
-// These buckets are allocated during deployment of the control plane.
-// If there are no free buckets the creation of the runtime environment for the application will fail.
-// TODO - In the future need to implement dynamic provisioning of buckets for implicit copy.
+// GetCopyDestination creates a Dataset for bucket allocation by implicit copies or ingest.
 func (m *ModuleManager) GetCopyDestination(item modules.DataInfo, destinationInterface *app.InterfaceDetails, geo string) (*app.DataStore, error) {
 	// provisioned storage for COPY
 	originalAssetName := item.DataDetails.Name
-	bucket := FindAvailableBucket(m.Client, m.Log, m.Owner, item.Context.DataSetID, originalAssetName, false, geo)
-	if bucket == nil {
-		return nil, errors.New(app.InsufficientStorage)
+	var dataset *comv1alpha1.Dataset = nil
+	var err error
+	if dataset, err = AllocateBucket(m.Client, m.Log, m.Owner, originalAssetName, geo); err != nil {
+		return nil, err
+	}
+	if err = m.Provision.CreateDataset(dataset); err != nil {
+		return nil, err
+	}
+	m.Datasets[item.Context.DataSetID] = dataset
+	bucketName := dataset.Spec.Local["bucket"]
+	secretRef := types.NamespacedName{Name: dataset.Spec.Local["secret-name"], Namespace: dataset.Spec.Local["secret-namespace"]}
+	if err = m.RegisterSecretInVault(bucketName, secretRef); err != nil {
+		return nil, err
 	}
 	connection, err := serde.ToRawExtension(&pb.DataStore{
 		Type: pb.DataStore_S3,
 		Name: "S3",
 		S3: &pb.S3DataStore{
-			Bucket:    bucket.Spec.Name,
-			Endpoint:  bucket.Spec.Endpoint,
-			ObjectKey: bucket.Status.AssetPrefixPerDataset[item.Context.DataSetID],
+			Bucket:    dataset.Spec.Local["bucket"],
+			Endpoint:  dataset.Spec.Local["endpoint"],
+			ObjectKey: originalAssetName + utils.Hash(originalAssetName, 10),
 		},
 	})
 	if err != nil {
 		return nil, err
 	}
 
 	return &app.DataStore{
-		CredentialLocation: utils.GetFullCredentialsPath(bucket.Spec.VaultPath),
+		CredentialLocation: utils.GetFullCredentialsPath(utils.GetDatasetVaultPath(bucketName)),
 		Connection:         *connection,
 		Format:             string(destinationInterface.DataFormat),
 	}, nil
 }
 
+func (m *ModuleManager) RegisterSecretInVault(id string, secretRef types.NamespacedName) error {
```

Comment:
We will probably need this one for other uses, can you generalize it and move out of module manager? (e.g., `pkg/vault`)

(can be in separate PR)

Reply:
OK, will do in  a new PR